numpy/_core/src/multiarray/nditer_pywrap.c

```diff
@@ -729,32 +717,60 @@ npyiter_init(NewNpyArrayIterObject *self, PyObject *args, PyObject *kwds)
         return -1;
     }
 
-    /* Set the dtypes and ops to all NULL to start */
-    memset(op_request_dtypes, 0, sizeof(op_request_dtypes));
+    /* Need nop to set up workspaces */
+    PyObject **op_objs = NULL;
+    PyObject *op_in_owned;  /* Sequence/object owning op_objs. */
+    int nop = npyiter_prepare_ops(op_in, &op_in_owned, &op_objs);
+    if (nop < 0) {
+        npy_free_cache_dim_obj(itershape);
+        return -1;
+    }
+
+    /* allocate workspace for Python objects (operands and dtypes) */
+    NPY_ALLOC_WORKSPACE(op, PyArrayObject *, 2 * 8, 2 * nop);
+    if (op == NULL) {
+        npy_free_cache_dim_obj(itershape);
+        Py_DECREF(op_in_owned);
+        return -1;
+    }
+    memset(op, 0, sizeof(PyObject *) * 2 * nop);
+    PyArray_Descr **op_request_dtypes = (PyArray_Descr **)(op + nop);
+
+    /* And other workspaces (that do not need to clean up their content) */
```

Comment:
If one put the loop over operands inside `finish` inside an `if (op != NULL)` then one could move all these together and simplify it here. (Really just nitpicking, trying the keep the main flow simple)